src/dist/client_auth.rs

```diff
@@ -486,44 +416,31 @@ mod implicit {
 
         Ok(response)
     }
+}
 
-    use super::*;
-    #[derive(Copy, Clone, Debug)]
-    pub struct Implicit;
-
-    impl hyper::service::Service<Request<Body>> for Implicit {
-        type Response = Response<Body>;
-        type Error = hyper::Error;
-        type Future = std::pin::Pin<
-            Box<
-                dyn 'static
-                    + Send
-                    + futures_03::Future<Output = result::Result<Self::Response, Self::Error>>,
-            >,
-        >;
-
-        fn poll_ready(
-            &mut self,
-            _cx: &mut task_03::Context<'_>,
-        ) -> task_03::Poll<result::Result<(), Self::Error>> {
-            task_03::Poll::Ready(Ok(()))
-        }
-
-        fn call(&mut self, req: Request<Body>) -> Self::Future {
-            let uri = req.uri().clone();
-            let fut = async move {
-                serve(req)
-                    .await
-                    .or_else(|e| super::error_code_response(uri, e))
-            };
-            Box::pin(fut)
-        }
-    }
+// Typing out a hyper service is a major pain, so let's focus on our simple
+// `fn(Request<Body>) -> Response<Body>` handler functions; to reduce repetition
+// we create a relevant service using hyper's own helper factory functions.
+macro_rules! make_service {
+    ($serve_fn: expr) => {{
+        use core::convert::Infallible;
+        use hyper::{Body, Request};
+        use hyper::server::conn::AddrStream;
+        use hyper::service::{make_service_fn, service_fn};
+
+        make_service_fn(|_socket: &AddrStream| async {
+            Ok::<_, Infallible>(service_fn(|req: Request<Body>| async move {
+                let uri = req.uri().clone();
+                $serve_fn(req)
+                    .or_else(|e| error_code_response(uri, e))
+            }))
+        })
+    }}
```

Comment:
To bring up some context, the way the `client_auth.rs` is structured now is that it separates 2 modules: `code_grant_pkce` and `implicit`, both of which have a state behind a `Mutex` in `lazy_static` and which expose a simple blocking/serial handler function `serve: fn(req: Request<Body>) -> Result<Response<Body>>)` along with other helper functions like `finish_url` or `code_grant_pkce::generate_verifier_challenge`. These are used in the 
- `get_token_oauth2_code_grant_pkce`
- `get_token_oauth2_implicit`

Currently, these functions spawn the server inside but the only thing we want from it is the `port` and for it to run with a graceful shutdown, so it's not strictly necessary to be able to name/refer to the specific `Service` implementation.

What comes to mind is rewriting the `code_grant_pkce` and `implicit` modules to be a dedicated `Service` implementer encapsulating the relevant state, meant to be running on a single thread/in a serial manner. This would make things less awkward around state locking and more explicit wrt requests being handled in a serial fashion. This would be even more type-safe, as it'd be harder to, let's say, mix up `implicit::serve` with `code_grant_pkce`-related functions but that seems like an bigger improvement/refactor that we could pursue in a follow-up.

Doing that would allow us to bring back `ServiceSpawner<S>` where `S` would be our new `Service` implementers if needed, but with the way things are structured now with dedicated modules and helper functions I'd prefer not to as it adds considerable complexity without giving us concrete benefits.

As such, I'd prefer to defer that and keep the change smaller for now to facilitate the migration itself - after that I'm open to actively improving what we have =)